torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -1360,6 +1373,39 @@ def _register_post_backward_hooks(
         flat_param._post_backward_hook_state = (acc_grad, hook_handle)  # type: ignore[attr-defined]
 
 
+def _register_post_backward_reshard_only_hooks(
```

Comment:
looks like now we have 2 paths where params can be resharded in post backward:
- for flat parameters that don't require grad, do it via a hook on the input activations
- for flat parameters that do require grad, do it with the standard post backward hook

could we unify and do both the reshards with the input activations hook?

Reply:
Resharding on the input activation can be later than our existing post-backward hook, which may regress memory. We should keep both paths and view this newly added path as the fallback for the `requires_grad=False` case.